### PR TITLE
Add lifecycle callbacks for workflow cache

### PR DESCRIPTION
**Pull request by stephanos**
## What changed?

Adds two lifecycle callbacks for workflow cache: OnPut and OnEvict.

## Why?

Prerequisite for https://github.com/temporalio/temporal/pull/6090

tl;dr We need to track workflow contexts per shard in order to clear them when the shard closes. This is the first step towards that goal.

## How did you test it?

No actual behavior change is introduced here yet. The callbacks have a new unit test.

## Potential risks

No new behavior, therefore no risk expected.

### CODE REVIEW

File: service/history/workflow/cache/cache.go [hunk 1]
```diff
@@ -147,19 +157,30 @@ func newCache(
...
-	opts.Pin = true
+	opts := &cache.Options{
+		TTL: ttl,
+		Pin: true,
+		OnPut: func(val any) {
+			// TODO: will be used in follow-up PR
+		},
+		OnEvict: func(val any) {
+			// TODO: will be used in follow-up PR
+		},
```
Comment (stephanos) [inline service/history/workflow/cache/cache.go:171]: This is where the callback has access to (1) the finalizer, (2) the workflow context and (3) the logger.

File: service/history/workflow/cache/cache.go [hunk 2]
```diff
@@ -84,13 +84,19 @@ type (
...
-	CacheImpl struct {
+	cacheImpl struct {
 		cache.Cache
 
+		onPut                     func(wfContext *workflow.Context)
+		onEvict                   func(wfContext *workflow.Context)
+		logger                    log.Logger
 		nonUserContextLockTimeout time.Duration
 	}
+	cacheItem struct {
```
Comment (alexshtin) [inline service/history/workflow/cache/cache.go:95]: Do you really need it now?

Comment (stephanos) [reply service/history/workflow/cache/cache.go:95]: Yes, it will contain the `Finalizer` in the next PR. The Finalizer cannot be passed into the cache since it has a different lifecycle.

File: service/history/workflow/cache/cache.go [hunk 3]
```diff
@@ -84,13 +84,19 @@ type (
 		) (workflow.Context, ReleaseCacheFunc, error)
 	}
 
-	CacheImpl struct {
+	cacheImpl struct {
```
[the remark this answers is not in the document]
Comment (stephanos) [reply service/history/workflow/cache/cache.go:87]: True!

File: common/cache/lru.go [hunk 4]
```diff
@@ -309,8 +314,9 @@ func (c *lru) putInternal(key interface{}, value interface{}, allowUpdate bool)
 	// If the entry exists, check if it has expired or update the value
 	if elt != nil {
 		existingEntry := elt.Value.(*entryImpl)
-		if !c.isEntryExpired(existingEntry, time.Now().UTC()) {
+		if !c.isEntryExpired(existingEntry, c.timeSource.Now().UTC()) {
```
Comment (stephanos) [inline common/cache/lru.go:317]: drive-by fix

File: service/history/workflow/cache/cache.go [hunk 5]
```diff
@@ -462,3 +487,10 @@ func makeCacheKey(
 		ShardUUID:   shardContext.GetOwner(),
 	}
 }
+
+func (c *cacheItem) CacheSize() int {
+	if sg, ok := c.wfContext.(cache.SizeGetter); ok {
+		return sg.CacheSize()
+	}
+	return 0
+}
```
Comment (stephanos) [inline service/history/workflow/cache/cache.go:496]: Redirects the cache size request to the inner workflow context.

Note that the cast is needed since some tests pass a MockContext in that does not implement this interface.

**Review by alexshtin — APPROVED** [review]
